### ISSUE TITLE
Optional extensions on byrefs are allowed to be declared but cannot be used

### ISSUE THREAD

**Issue by cartermp** [Microsoft/visualfsharp#5354]
Consider the following:

```fsharp
type byref<'T> with
    member __.Foo() = ()

let x = 12
let y = &x
y.Foo()
```

This will fail to compile:

`
The field, constructor, or member 'Foo' is not defined.
`

However, it seems allowable to declare `Foo`, and IntelliSense also suggests it:

![image](https://user-images.githubusercontent.com/6309070/42972567-c2046946-8b64-11e8-9b08-83767f74b5c2.png)

I think the declaration should not be allowed if it cannot be used.

**Note:** Using a "floating in space" C#-style extension method works, which will further add to the confusion surrounding type extensions, but that's a bit orthogonal.

**Comment by TIHan** [comment]
I don't know if this is a bug in the compiler, but a possible hole in the spec as it doesn't mention this type of extensibility for byrefs.

**Comment by cartermp** [comment]
I'll defer to @dsyme on if optional extensions should be supported for byrefs, but either way there is work that will need to be done in the compiler:

* Allow them, thus fixing the call site error
* Disallow them, meaning we'll need to emit an error at the declaration site

**Comment by zpodlovics** [comment]
As Span already target only the experienced F# developers to do performance critical work in F# (instead of rewriting everything performance critical in C#). Please support Span to be on par with C# both for F# consumption (C# -> F#) both for production (F# -> C#).

**Comment by cartermp** [comment]
@zpodlovics do you have a specific concern w.r.t optional type extensions and `byref`/`inref`/`outref`?

**Comment by cartermp** [comment]
@zpodlovics You can do "floating in space" extensions on byrefs today just fine:

```fsharp
open System.Runtime.CompilerServices
[<Extension>]
type Ext = 
    [<Extension>]
    static member ExtDateTime2(dt: inref<DateTime>, x:int) = dt.AddDays(double x)
```

These are equivalent to C# extension methods and are guaranteed to work as per the Span spec.

The question is if F# optional type extensions are possible as well. These are entirely different things, and though they are more "natural" to F#, there may be issues with them that could prevent their usage.

**Comment by zpodlovics** [comment]
@cartermp You are right, your C# compat example should also work with byreflike types too. However it would be good to have consistency for the F# optional type extensions. Surprisingly even the ```nativeptr<'T>``` works with F# optional type extensions.

These examples works as expected:

```fsharp
type array1<'T> = 
  val Value: 'T
  new(v) = { Value=v }

type array1<'T> with
  member this.Foo() = printfn "foo"

let x = array1<int>(1)
x.Foo()
```

```fsharp
type array2<'T when 'T: unmanaged> =              
  val Value: 'T                                   
  new(v) = { Value=v }   

type array2<'T when 'T: unmanaged> with           
  member this.Foo() = printfn "foo"

let x = array2<int>(1)
x.Foo()
```

```fsharp
type nativeptr<'T when 'T: unmanaged> with
  member this.Foo() = printfn "foo"

let x = 0n |> NativeInterop.NativePtr.ofNativeInt<int>
x.Foo()

```

Type aliases will not work:

```fsharp
type nptr<'T when 'T: unmanaged> = nativeptr<'T>
```

```fsharp
type nptr<'T> with
  member this.Foo() = printfn "foo"

```

```
error FS0957: One or more of the declared type parameters for this type extension have a missing or wrong type constraint not matching the original type constraints on 'nptr<_>'
```

```fsharp
type nptr<'T when 'T: unmanaged> with
  member this.Foo() = printfn "foo"
```

```
error FS0957: One or more of the declared type parameters for this type extension have a missing or wrong type constraint not matching the original type constraints on 'nptr<_>'
```

```
error FS0964: Type abbreviations cannot have augmentations
```

is there any way (from F# code) to check whether a type name is a real F# type or just a type alias ?

**Comment by cartermp** [comment]
@zpodlovics type abbreviations will not work by design - this is just a rule for optional type extensions. Aside from editor tooling (where the real type name is shown), I'm not aware of another way to check.

For example, FSI is actually eliding information:

```
> type i = System.Int32;;
type i = System.Int32

> let f (x: i) = x + 1;;
val f : x:i -> i

> f 12;;
val it : i = 13
```

This is an orthogonal concern, but should be tracked in [fslang-suggestions](https://github.com/fsharp/fslang-suggestions) since there are some design questions to hash out about how type abbreviations can better show their information in F# interactive.

**Comment by cartermp** [comment]
Fixed by #5472
